### PR TITLE
[REVIEW] Add a minimal suite of nvstrings and nvcategory pytests

### LINKED ISSUE #266: [FEA] Add a minimal suite of nvstrings pytests
We need to add pytests for nvstrings functionality to help developers make changes to existing functionality without worrying about breaking other functionality (avoid unintentional build failures).

### PR DESCRIPTION
**Pull request by beckernick**
**Summary of Changes**
- This PR adds a minimal suite of nvstrings and nvcategory pytests based on the existing print statements
- Removes unnecessary test files if they have been subsumed by tests in other files

Basic Test Suite

### **nvStrings**

- [x] add_strings
- [x] byte_count
- [x] capitalize
- [x] cat
- [x] center
- [x] compare
- [x] contains
- [x] copy
- [x] count
- [x] endswith
- [x] extract
- [x] extract_record
- [x] fillna
- [x] find
- [x] find_from
- [x] find_multiple
- [x] findall
- [x] findall_record
- [x] gather
- [x] get
- [ ] get_cpointer
- [x] hash
- [x] htoi
- [x] index
- [x] ip2int
- [x] isalnum
- [x] isalpha
- [x] isdecimal
- [x] isdigit
- [x] islower
- [x] isnumeric
- [x] isspace
- [x] isupper
- [x] join
- [x] len
- [x] ljust
- [x] lower
- [x] lstrip
- [ ] m_cptr
- [x] match
- [x] null_count
- [x] order
- [x] pad
- [x] partition
- [x] remove_strings
- [x] repeat
- [x] replace
- [x] replace_with_backrefs
- [x] rfind
- [x] rindex
- [x] rjust
- [x] rpartition
- [x] rsplit
- [x] rsplit_record
- [x] rstrip
- [ ] set_null_bitmask
- [x] size
- [x] slice
- [x] slice_from
- [x] slice_replace
- [x] sort
- [x] split
- [x] split_record
- [x] startswith
- [x] stof
- [x] stoi
- [x] strip
- [x] sublist
- [x] swapcase
- [x] title
- [x] to_host
- [x] to_offsets
- [x] translate
- [x] upper
- [x] wrap
- [x] zfill
- [ ] bind_cpointer
- [x] free
- [x] from_csv
- [x] from_offsets
- [x] from_strings
- [x] int2ip
- [x] itos
- [x] to_device
- [x] from_booleans
- [x] to_booleans
- [ ] get_ipc_data
- [ ] create_from_ipc
- [x] is_empty
- [x] match_strings
- [x] stod
- [x] stol
- [x] timestamp2int
- [x] dtos
- [x] ftos
- [x] int2timestamp
- [x] ltos

### **nvcategory**

- [x] add_keys
- [x] add_strings
- [x] gather
- [x] gather_and_remap
- [x] gather_strings
- [ ] get_cpointer
- [x] indexes_for_key
- [x] keys
- [x] keys_size
- [ ] m_cptr
- [x] merge_and_remap
- [x] merge_category
- [x] remove_keys
- [x] remove_strings
- [x] remove_unused_keys
- [x] set_keys
- [x] size
- [x] to_strings
- [x] value
- [x] value_for_index
- [x] values
- [ ] values_cpointer
- [ ] bind_cpointer
- [x] from_offsets
- [x] from_strings
- [x] from_strings_list
- [x] to_device

This closes #266

### CODE REVIEW

**Comment by beckernick** [comment]
Moving this to review. @davidwendt , @randerzander let me know what you think of this as an initial start It's intended to be comprehensive in terms of method coverage, but not in terms of edge cases.

File: python/tests/test_translate.py [hunk 1]
```diff
@@ -1,27 +1,55 @@
...
-strs = nvstrings.to_device(["hello","there","world","accéntéd",None,""])
-print(strs)
-print(".translate():",strs.translate([]))
-print(".translate([[e,a]]):",strs.translate([['e','a']]))
-print(".translate([[e,é]]):",strs.translate([['e','é']]))
-print(".translate([[é,e],[o,None]]):",strs.translate([['é','e'],['o',None]]))
 
-print(".translate(maketrans(e,a):",strs.translate(str.maketrans('e','a')))
-print(".translate(maketrans(elh,ELH):",strs.translate(str.maketrans('elh','ELH')))
+@pytest.mark.parametrize('table', [
```
Comment (davidwendt) [inline python/tests/test_translate.py]: What is happening here? Is Pandas required to verify this translate?

**Comment by davidwendt** [comment]
Forgot to say that this looks awesome Nick.

File: python/tests/test_translate.py (hunk 1 above)
Comment (beckernick) [reply python/tests/test_translate.py]: Pandas is not required, but in some areas may be desirable for consistency with the `series.str` API. No individual test inherently needs to be comparing with pandas. We could compare this to the Python translate API, which would also require an ordinal mapping

**Comment by davidwendt** [comment]
The `test_column.py` could be replaced with a `test_csv.py` perhaps.
Also, `test_substr_tw.py` is probably not necessary to preserve.

**Comment by beckernick** [comment]
We can add tests for RMM compatibility. I was modeling after cuDF python layer unit tests which don't separately test RMM compatibility as far as I can tell.

**Comment by davidwendt** [comment]
> We can add tests for RMM compatibility. I was modeling after cuDF python layer unit tests which don't separately test RMM compatibility as far as I can tell.

I'm more concerned the code works with RMM enabled than without.

File: python/tests/test_translate.py (hunk 1 above)
Comment (davidwendt) [reply python/tests/test_translate.py]: You should be able to use the same translate-table for both nvstrings and pandas:
```
import pandas as pd
import nvstrings
ttbl = str.maketrans('e','a')
ss = ['hello','there','world','accéntéd',None,'']
ps = pd.Series(ss)
ds = nvstrings.to_device(ss)
print(ps.str.translate(ttbl))
print(ds.translate(ttbl))
```
```
0       hallo
1       thara
2       world
3    accéntéd
4        None
5            
dtype: object

['hallo', 'thara', 'world', 'accéntéd', None, '']
```
If that helps.

Comment (beckernick) [reply python/tests/test_translate.py]: This does help, thanks! The reason I have them separate is that the pandas method exclusively accepts an ordinal:character mapping dictionary (such as from the translate table), but the nvstrings method does not. It can accept the mapping as either a list of a dictionary. In the ordinal test, its doing what you've described with the `str.maketrans('elh', 'ELH'),` parameter to the pytest. Let me know if you think we should change/combine them!

**Comment by beckernick** [comment]
@davidwendt , all tests now use RMM

**Comment by beckernick** [comment]
@randerzander @raydouglass should we not be using `librmm_cffi` anymore given the ongoing Cythonization in cuDF? CI is failing due to the `from librmm_cffi import librmm as rmm` import, which has generally been the [standard way](https://github.com/rapidsai/rmm#handling-rmm-options-in-python-code) of importing RMM

**Comment by beckernick** [comment]
Looks like the librmm_cffi issue is packaging related based on the discussion in rapidsai/dask-cudf#215

**Comment by beckernick** [comment]
rerun tests

**Comment by beckernick** [comment]
Still getting the module not found error on `librmm_cffi`, despite the rapidsai/dask-cudf#215 appearing to be past it. @rlratzel , is there any action needed here?

File: python/tests/test_datetime.py [hunk 2]
```diff
@@ -1,31 +1,33 @@
...
-rmm_cfg.use_pool_allocator = True
-rmm.initialize()
-
-s = nvstrings.to_device(["2019-03-20T12:34:56Z","2020-02-29T23:59:59Z"])
-print(s)
-print(".timestamp2int()",s.timestamp2int())
-print(".timestamp2int(ms)",s.timestamp2int(units='ms'))
-
-s = nvstrings.to_device(["04/24/2019","02/29/2020"])
-print(".timestamp2int(M/D/Y)",s.timestamp2int(format='%m/%d/%Y'))
```
Comment (davidwendt) [inline python/tests/test_datetime.py]: Seems we are missing a bunch of testcases here.

Comment (beckernick) [reply python/tests/test_datetime.py]: Agreed, this is a minimal test case. We absolutely will want to create tests for the other formats noted in `test_datetime`.

**Comment by beckernick** [comment]
We can add some additional whitespace examples. I took the whitespace array and added a couple of those instances to existing tests (at the end)

```
s = ["héllo", None, "a_bc_déf", "a__bc", "_ab_cd", "ab_cd_", "", ' a b ',
         ' a  bbb   c']
```

Let me know if you think we should include more for this initial set of minimal tests.

I did remove `test_from.py`. Tests  `find_from` and `slice_from` should be in test_compare and test_substr.

**Comment by raydouglass** [comment]
> Still getting the module not found error on `librmm_cffi`, despite [rapidsai/dask-cudf#215](https://github.com/rapidsai/dask-cudf/pull/215) appearing to be past it. @rlratzel , is there any action needed in this repository separately?

This should be resolved now. rerun tests

**Comment by beckernick** [comment]
rerun tests

**Comment by beckernick** [comment]
rerun tests

File: python/tests/test_translate.py (hunk 1 above)
Comment (beckernick) [reply python/tests/test_translate.py]: @davidwendt given our Slack conversation, are we good to consider this resolved for now with the plan to transition away in the future?

File: python/tests/test_datetime.py (hunk 2 above)
Comment (beckernick) [reply python/tests/test_datetime.py]: Are we comfortable with only the minimal format test for this initial PR, or do we need to expand a bit?

File: python/tests/test_translate.py (hunk 1 above)
Comment (davidwendt) [reply python/tests/test_translate.py]: Yes
